x/stakeibc/keeper/msg_server_claim_undelegated_tokens.go

```diff
@@ -4,75 +4,75 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/Stride-Labs/stride/utils"
+	recordstypes "github.com/Stride-Labs/stride/x/records/types"
+
 	"github.com/Stride-Labs/stride/x/stakeibc/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
 	bankTypes "github.com/cosmos/cosmos-sdk/x/bank/types"
 )
 
-func min(a int, b int) int {
-	if a < b {
-		return a
-	}
-	return b
-}
+var min = utils.Min
 
 func (k msgServer) ClaimUndelegatedTokens(goCtx context.Context, msg *types.MsgClaimUndelegatedTokens) (*types.MsgClaimUndelegatedTokensResponse, error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
-	// grab our host zone
-	hostZone, found := k.GetHostZone(ctx, msg.HostZone)
+
+	// grab the UserRedemptionRecord from the store
+	userRedemptionRecordKey := recordstypes.UserRedemptionRecordKeyFormatter(msg.HostZoneId, msg.Epoch, msg.Sender)
+	userRedemptionRecord, found := k.RecordsKeeper.GetUserRedemptionRecord(ctx, userRedemptionRecordKey)
 	if !found {
-		errMsg := fmt.Sprintf("Host zone %s not found", msg.HostZone)
+		errMsg := fmt.Sprintf("User redemption record %s not found on host zone %s", userRedemptionRecordKey, msg.HostZoneId)
 		k.Logger(ctx).Error(errMsg)
-		return nil, sdkerrors.Wrap(types.ErrInvalidHostZone, errMsg)
+		return nil, sdkerrors.Wrapf(types.ErrInvalidUserRedemptionRecord, "could not get user redemption record: %s", userRedemptionRecordKey)
+	}
+
+	// check that the record is claimable
+	if !userRedemptionRecord.GetIsClaimable() {
+		errMsg := fmt.Sprintf("User redemption record %s is not claimable", userRedemptionRecord.Id)
+		k.Logger(ctx).Error(errMsg)
+		return nil, sdkerrors.Wrapf(types.ErrInvalidUserRedemptionRecord, "user redemption record is not claimable: %s", userRedemptionRecordKey)
 	}
+
 	// grab necessary fields to construct ICA call
-	var msgs []sdk.Msg
-	redemptionAccount, found := k.GetRedemptionAccount(ctx, hostZone)
+	hostZone, found := k.GetHostZone(ctx, msg.HostZoneId)
```

Comment:
Let's move this check to the top

Reply:
Can you elaborate on this?

Reply:
part of my PR that does validation on inputs was to check the validity of hostzones in each msg_server that has a hostZone in its msg (like this msg_server). I always do that at the top of the msg_server. Mostly for consistency

Reply:
got it - I think this should still fail if the host zone doesn't exist! there isn't any substantive logic above it, just checks on UserRedemptionRecords. slight preference to keep this here so the sections of the function are
1. Checks on UserRedemptionRecord
2. Checks on all fields required for the ICA
3. Constructing and sending the transaction, updating state